core/drivers/stm32_rtc.c

```diff
@@ -460,6 +460,95 @@ static void stm32_rtc_to_tm(uint32_t ssr, uint32_t tr, uint32_t dr,
 		      ((dr & RTC_DR_YU_MASK) >> RTC_DR_YU_SHIFT) + YEAR_REF;
 }
 
+static TEE_Result stm32_rtc_init(void)
+{
+	uint32_t pred_a_max = RTC_PRER_PREDIV_A_MASK >> RTC_PRER_PREDIV_A_SHIFT;
+	uint32_t pred_s_max = RTC_PRER_PREDIV_S_MASK >> RTC_PRER_PREDIV_S_SHIFT;
+	unsigned long rate = clk_get_rate(rtc_dev.rtc_ck);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	vaddr_t base = get_base();
+	uint32_t pred_a = 0;
+	uint32_t pred_s = 0;
+	uint32_t prer = io_read32(base + RTC_PRER);
+	uint32_t cr = io_read32(base + RTC_CR);
+
+	if (rate > (pred_a_max + 1) * (pred_s_max + 1))
+		panic("rtc_ck rate is too high");
+
+	if (cr & RTC_CR_FMT && !IS_ENABLED(CFG_STM32_RTC_HIGH_ACCURACY))
+		return TEE_SUCCESS;
+
+	if (IS_ENABLED(CFG_STM32_RTC_HIGH_ACCURACY)) {
+		/*
+		 * Compute the prescaler values whom divides the clock in order
+		 * to get a * 1 Hz output by maximizing accuracy
+		 * (maximizing PREDIV_S).
+		 */
+		for (pred_a = 0; pred_a <= pred_a_max; pred_a++) {
+			pred_s = (rate / (pred_a + 1)) - 1;
+			if (pred_s <= pred_s_max &&
+			    ((pred_s + 1) * (pred_a + 1)) == rate)
+				break;
+		}
+
+		/*
+		 * 1 Hz output not possible, give priority to RTC power
+		 * consumption by choosing the higher possible value for
+		 * prediv_a
+		 */
+		if (pred_s > pred_s_max || pred_a > pred_a_max) {
+			pred_a = pred_a_max;
+			pred_s = (rate / (pred_a + 1)) - 1;
+
+			DMSG("rtc_ck is %s",
+			     (rate < ((pred_a + 1) * (pred_s + 1))) ?
+			     "fast" : "slow");
+		}
+
+		prer &= RTC_PRER_PREDIV_S_MASK | RTC_PRER_PREDIV_A_MASK;
+		pred_s = SHIFT_U32(pred_s, RTC_PRER_PREDIV_S_SHIFT) &
+			 RTC_PRER_PREDIV_S_MASK;
+		pred_a = SHIFT_U32(pred_a, RTC_PRER_PREDIV_A_SHIFT) &
+			 RTC_PRER_PREDIV_A_MASK;
+	}
+
+	/* Return if there is nothing to initialize */
+	if (IS_ENABLED(CFG_STM32_RTC_HIGH_ACCURACY) &&
+	    prer == (pred_s | pred_a))
+		return TEE_SUCCESS;
+
+	stm32_rtc_write_unprotect();
+
+	res = stm32_rtc_enter_init_mode();
+	if (res) {
+		EMSG("Can't enter in init mode. Fail to initialize RTC.");
+		stm32_rtc_write_protect();
+		return TEE_ERROR_BUSY;
+	}
+
+	if (IS_ENABLED(CFG_STM32_RTC_HIGH_ACCURACY)) {
+		io_write32(base + RTC_PRER, pred_s);
+		io_write32(base + RTC_PRER, pred_a | pred_s);
+	}
+
+	/* Force 24h time format */
+	cr &= ~RTC_CR_FMT;
+	io_write32(base + RTC_CR, cr);
+
+	stm32_rtc_exit_init_mode();
+
+	res = stm32_rtc_wait_sync();
+	if (res) {
+		EMSG("Can't sync RTC. Fail to initialize RTC.");
+		stm32_rtc_write_protect();
+		return TEE_ERROR_BUSY;
```

Comment:
```suggestion
		res = TEE_ERROR_BUSY;
```
and `return res;` below.

By the way, `stm32_rtc_exit_init_mode();` is always followed by `stm32_rtc_wait_sync()`.
The 2 functions could be merged, the same way `stm32_rtc_enter_init_mode()` already waits RTC expected state.